### PR TITLE
[Data] feat: add streaming train test split implementation

### LINKED ISSUE #56780: [Data]: Streaming train/test split
### Description

The current implementation of `train_test_split` materialized the datasets, which for very large datasets leads to bottlenecks or excesively big cluster to be able to materialize them. 
Adding support for a streaming version of the spilt would allow for less resource consumption.

There's a few ways that this can be implemented:

## Bernoulli split
- Each record goes to A with probability p, else to B.
- One pass, no buffering, truly streaming.
- Result sizes will be close to p:N but not exact.

## Hash-based split
- Choose a stable key (e.g., user_id). Hash it; if hash < threshold, send to A; else B.
- Proportions are very close to p for large datasets, and it’s reproducible (no RNG state needed).
- Works great for training/validation splits that must be consistent across runs.

## Exact-size split in one pass (requires knowing N)
- If you know total N and want exactly k = round(p*N) in A, you can do selection sampling:
  - As you read item i, with n_remaining = N - i and k_remaining, choose current item for A with probability k_remaining / n_remaining.
 -  Guarantees exactly k in A, uniformly at random, single pass, O(1) memory.

For our use case `Hash based split` would be the ideal given it guarantees determinism, though `Bernoulli` will work for a wide range of use cases, and does not require a unique key in the dataset

### Use case

Splitting input dataset into train/test splits for training/validation purposes.

### PR DESCRIPTION
**Pull request by martinbomio**
## Why are these changes needed?

To allow for a streaming implementation of train/test split

## Related issue number

Closes https://github.com/ray-project/ray/issues/56780

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---

> [!NOTE]
> Adds Dataset.streaming_train_test_split for streaming train/test splits (random or hash-based) with validations and tests.
> 
> - **API (Dataset)**:
>   - Add `Dataset.streaming_train_test_split(test_size, split_type="random"|"hash", hash_column=None, seed=None, **ray_remote_kwargs)` for streaming train/test splits.
>     - Implements per-batch bucketing via `map_batches` and splits via filtering; supports deterministic random seeding per task and hash-based partitioning on `hash_column`.
>     - Validates inputs (`test_size` in (0,1), required `hash_column` for hash, disallow `seed` with hash, etc.).
>     - Returns two `Dataset`s (`train`, `test`).
> - **Tests**:
>   - Add large-scale tests for hash and random splits verifying proportions and disjointness.
>   - Add param validation tests covering invalid `test_size`, `split_type`, `hash_column`, and `seed` combinations.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 24f678c582428f65500eeef0ee4fb774fa045ee8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -2463,6 +2463,125 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
             )
         return ds_length
 
+    @ConsumptionAPI
```
Comment (richardliaw) [inline python/ray/data/dataset.py]: nit, this isn't a consumption API

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -2463,6 +2463,124 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+        This method is recommended for large datasets.
+
+        The split type can be either "hash" or "bernoulli".
+        - "hash": The dataset is split into train and test subsets based on the hash of the key column.
+        - "bernoulli": The dataset is split into train and test subsets based on the Bernoulli distribution.
+
+        Examples with Bernoulli split:
+
+            >>> import ray
+            >>> ds = ray.data.range(8, override_num_blocks=1)
```
Comment (martinbomio) [inline python/ray/data/dataset.py]: I had to override the num_blocks when the data is very small, otherwise it will create one block per record and the `bernulli` split will put all records into the same bucket

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -2463,6 +2463,126 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+
+        def bernoulli_split(batch, test_proportion: float, seed: int | None):
+            # Build a stable per-task seed so re-execution gives identical masks
+            base = 0 if seed is None else seed
+            task_id_str = ray.get_runtime_context().get_task_id()
+            # task_id is stable for a given execution of the upstream block
+            task_bits = int(task_id_str[-16:], 16) & 0xFFFFFFFF
+
+            rng = np.random.default_rng(base ^ task_bits)
+            is_train = rng.random(batch.num_rows) < (1 - test_proportion)
```
Comment (goutamvenkat-anyscale) [inline python/ray/data/dataset.py]: Can you please add some comments to elaborate on what's going on here? 
An example would be helpful.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -2463,6 +2463,126 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
             )
         return ds_length
 
+    @PublicAPI(stability="alpha", pi_group=SMJ_API_GROUP)
+    def streaming_train_test_split(
```
Comment (goutamvenkat-anyscale) [inline python/ray/data/dataset.py:2467]: For my own understanding:
What was the rationale for having a separate API for streaming? Was introducing a streaming bool flag and split_type not sufficient in the existing `train_test_split` API?

Comment (martinbomio) [reply python/ray/data/dataset.py:2467]: I didn't see any of the other APIs to be extended this way, so thought it would be better to create a new one. I also didn't want to support everything that is supported in the `train_test_split` implementation, at least not on the first version of it.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (martinbomio) [reply python/ray/data/dataset.py]: @goutamvenkat-anyscale added comments and more docs in https://github.com/ray-project/ray/pull/56803/commits/68d14bd39ab4902a49d1e2542484cad3c5b010d3 PTAL!

File: python/ray/data/dataset.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/data/dataset.py:2467]: I also thought more about this - the other API is public/stable and has a ConsumptionAPI tag + the kwargs are noncomposable, so it made sense to separate it for now

File: python/ray/data/dataset.py [hunk 5]
```diff
@@ -2463,6 +2463,140 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+            if key_column in batch.column_names:
+                # Use provided key for hashing
+                keys = batch[key_column].to_numpy()
+            else:
+                raise ValueError(f"Key column {key_column} not found in batch")
+
+            bucket_arr = pa.array([key_to_bucket(key) for key in keys], type=pa.bool_())
+            return batch.append_column(_TRAIN_TEST_SPLIT_COLUMN, bucket_arr)
+
+        def bernoulli_split(batch, test_proportion: float, seed: Optional[int]):
```
Comment (goutamvenkat-anyscale) [inline python/ray/data/dataset.py]: Nit: missing type annotation

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -2463,6 +2463,140 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+
+            This version ensures that the random choices are **stable per Ray task execution** by seeding
+            the RNG with a combination of a user-specified seed and the Ray task ID.
+            """
+            # Build a stable per-task seed so re-execution gives identical masks
+            base = 0 if seed is None else seed
+            # Each Ray task that processes a batch has a unique, stable ID (UUID-like).
+            # We grab the last 16 hex digits of the task_id to create a 32-bit integer.
+            # This ensures *different tasks* don't reuse the same RNG sequence.
+            task_id_str = ray.get_runtime_context().get_task_id()
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: As I recall, `map_batches()` may reuse the Ray tasks. So different batches (of the same block) may get the same seed here, which should be easy to test.. In `random_sample()` we initialize the Generator once per task and store it in a `TaskContext`.

Comment (martinbomio) [reply python/ray/data/dataset.py]: ah I see, so you are saying that is better to just save the rng and simplify this logic?

Comment (martinbomio) [reply python/ray/data/dataset.py]: yeah that works pretty nicely! thank you!

**Comment by martinbomio** [comment]
@goutamvenkat-anyscale @wingkitlee0 mind taking another look?

File: python/ray/data/dataset.py [hunk 7]
```diff
@@ -2463,6 +2463,141 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
             )
         return ds_length
 
+    @PublicAPI(stability="alpha", pi_group=SMJ_API_GROUP)
+    def streaming_train_test_split(
+        self,
+        test_proportion: float,
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: I suggest we use `test_size` to follow `train_test_split` API.

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -2463,6 +2463,141 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+        key_column: Optional[str] = None,
+        seed: Optional[int] = None,
+        **ray_remote_kwargs,
+    ) -> Tuple["Dataset", "Dataset"]:
+        """split the dataset into train and test subsets in a streaming manner.
+        This method is recommended for large datasets.
+
+        The split type can be either "hash" or "bernoulli".
+        - "hash": The dataset is split into train and test subsets based on the hash of the key column.
+        - "bernoulli": The dataset is split into train and test subsets based on the Bernoulli distribution.
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: I suggest putting Bernoulli distribution first as it's the default. We could add that this is the simple per-row random selection (in case people didn't use the name Bernoulli, like myself..)

File: python/ray/data/dataset.py [hunk 9]
```diff
@@ -2463,6 +2463,141 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
             )
         return ds_length
 
+    @PublicAPI(stability="alpha", pi_group=SMJ_API_GROUP)
+    def streaming_train_test_split(
+        self,
+        test_proportion: float,
+        *,
+        split_type: Literal["hash", "bernoulli"] = "bernoulli",
+        key_column: Optional[str] = None,
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: So this `key_column` has nothing to do with `stratify` (which is in the `train_test_split` API). Do you see this API can be extended to stratifying split as well? It seems to be do-able with a groupby.

I suggest we call this `hash_column` or `split_column` to distinguish from `key`, which is commonly used in other APIs (`groupby`, `repartition`). It's also more obvious to not use it with "bernoulli" mode..

Later if we add `stratify` arg we do not need to deal with the confusion of `stratify` vs `key_column`.

Thoughts?

File: python/ray/data/dataset.py (hunk 8 above)
Comment (richardliaw) [reply python/ray/data/dataset.py]: yeah maybe not bernoulli but just random

File: python/ray/data/dataset.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/data/dataset.py]: agree

File: python/ray/data/dataset.py (hunk 8 above)
Comment (martinbomio) [reply python/ray/data/dataset.py]: done

File: python/ray/data/dataset.py (hunk 9 above)
Comment (martinbomio) [reply python/ray/data/dataset.py]: done,

**Comment by martinbomio** [comment]
@richardliaw @wingkitlee0 I am seeing an error on the build on docs, which look unrelated. Mind taking a quick look

File: python/ray/data/dataset.py [hunk 10]
```diff
@@ -2463,6 +2463,142 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+            >>> test.take_batch()
+            {'id': array([1, 7])}
+
+        Args:
+            test_size: The proportion of the dataset to include in the test split.
+                Must be between 0.0 and 1.0.
+            split_type: The type of split to perform. Can be "hash" or "random".
+            hash_column: The column to use for the hash split. Required for hash split and
+                ignored for Bernoulli split.
+            seed: The seed to use for the Bernoulli split. Ignored for hash split.
```
Comment (goutamvenkat-anyscale) [inline python/ray/data/dataset.py]: Docs should be updated to say random instead of Bernouilli

File: python/ray/data/tests/test_split.py [hunk 11]
```diff
@@ -887,6 +887,61 @@ def verify_integrity(splits):
...
+    ds = ray.data.range(10000000, override_num_blocks=10)
+
+    ds_train, ds_test = ds.streaming_train_test_split(
+        test_proportion=0.2, split_type="hash", key_column="id"
+    )
+
+    np.testing.assert_almost_equal(float(ds_train.count()) / 10000000.0, 0.8, decimal=3)
+    np.testing.assert_almost_equal(float(ds_test.count()) / 10000000.0, 0.2, decimal=3)
+
+    # Check if train and test are disjoint
```
Comment (srinathk10) [inline python/ray/data/tests/test_split.py:903]: Could we also check the actual contents of train and test dataset in addition to count and join?

File: python/ray/data/dataset.py [hunk 12]
```diff
@@ -2463,6 +2463,142 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+        if test_size < 0 or test_size > 1:
+            raise ValueError("Test proportion must be between 0 and 1.")
+
+        if seed is not None and split_type == "hash":
+            raise ValueError("seed is not supported for hash split")
+
+        if hash_column is not None and split_type == "random":
+            raise ValueError("hash_column is not supported for random split")
+
+        def random_split(batch: pa.Table):
```
Comment (srinathk10) [inline python/ray/data/dataset.py:2544]: When seed is provided, we would still need to ground this with key column, similar to hash for determinism?

File: python/ray/data/tests/test_split.py [hunk 13]
```diff
@@ -887,6 +887,61 @@ def verify_integrity(splits):
...
+    ds = ray.data.range(10000000, override_num_blocks=10)
+
+    ds_train, ds_test = ds.streaming_train_test_split(
+        test_size=0.2, split_type="random", seed=seed
+    )
+
+    np.testing.assert_almost_equal(float(ds_train.count()) / 10000000.0, 0.8, decimal=3)
+    np.testing.assert_almost_equal(float(ds_test.count()) / 10000000.0, 0.2, decimal=3)
+
+    # Check if train and test are disjoint
```
Comment (srinathk10) [inline python/ray/data/tests/test_split.py:918]: Same as above, with a seed, need to validate contents of train and test dataset.

File: python/ray/data/dataset.py (hunk 12 above)
Comment (martinbomio) [reply python/ray/data/dataset.py:2544]: can you expand on this? it doesn't seem necessary to do this and it will require a unique key, which is not always available

File: python/ray/data/tests/test_split.py (hunk 11 above)
Comment (martinbomio) [reply python/ray/data/tests/test_split.py:903]: are you saying that we should add an assertion for exact record match? is this really needed? we do not care about where each record is place as long as the proportion is respected and the sets are disjoint

File: python/ray/data/dataset.py (hunk 12 above)
Comment (richardliaw) [reply python/ray/data/dataset.py:2544]: Yeah I don't think you need to do that. You can seed to generate indices.

Comment (srinathk10) [reply python/ray/data/dataset.py:2544]: @martinbomio We can use [preserve_order](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/execution/interfaces/execution_options.py#L288) flag in ExecutionOptions to guarantee determinism here.

File: python/ray/data/dataset.py [hunk 14]
```diff
@@ -2463,6 +2463,146 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
             )
         return ds_length
 
+    @PublicAPI(stability="alpha", pi_group=SMJ_API_GROUP)
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: Typo for `api_group`. 

This should fix the doc and show this API in the shuffling section.

File: python/ray/data/dataset.py [hunk 15]
```diff
@@ -2463,6 +2463,146 @@ def _validate_test_size_int(self, test_size: int, ds: "Dataset") -> int:
...
+        **ray_remote_kwargs,
+    ) -> Tuple["Dataset", "Dataset"]:
+        """split the dataset into train and test subsets in a streaming manner.
+        This method is recommended for large datasets.
+
+        The split type can be either "hash" or "random".
+        - "random": The dataset is split into random train and test subsets.
+        - "hash": The dataset is split into train and test subsets based on the hash of the key column.
+
+        Important: Make sure to set the `preserve_order` flag in the `ExecutionOptions` to True
```
Comment (wingkitlee0) [inline python/ray/data/dataset.py]: optional. you can add `.. note::` or `..tip::` to highlight this. See `map_batches`.

**Review by wingkitlee0 — APPROVED** [review]
Looks great.

**Review by goutamvenkat-anyscale — APPROVED** [review]

**Comment by martinbomio** [comment]
@richardliaw build is finally green
